lib/spack/docs/build_systems/gopackage.rst

```diff
@@ -0,0 +1,164 @@
+
+to invoke ``go build -p 12``.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Slightly more complicated packages
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Packages that can be built with a simple invocation of ``go build``,
+that use modules, *but have not vendored their dependencies* require a
+bit of extra work.  In particular, in order to satisfy Spack's
+requirement that nothing is downloaded during the build step, the
+package needs to define "resources" that describe from whence to fetch
+each dependency, where to emplace it, and the version of the package
+to which it applies.  It is possible (likely) that each release of the
+project will new set of resources definitions covering the correct
+versions.
+
+In the general case, compiling these resource is complicated.
+
+Fortunately, packagers of projects that use modules can use an
+automated tool (``modules2tuple``) to provide the resource definitions
+and ``import_resources`` to include them without overly-complicating
+the package definition.  ``modules2tuple``
+(https://github.com/dmgk/modules2tuple) was originally developed by
+the FreeBSD Ports team to generate the Makefile fragment their build
+system uses to represent a Go application's dependencies.  For the
+curious, their Porters Handbook includes more details about `packaging
+Go Applications
+<https://www.freebsd.org/doc/en_US.ISO8859-1/books/porters-handbook/building.html#using-go>`_
+for FreeBSD.
+
+There are three steps to describing the dependencies in the package
+definition:
+
+* First, stage the application and use the Go tool chain to
+  solve the set of required dependencies and build a vendor directory.
+
+* Second, run the newly created ``vendor/modules.txt`` file through
+  ``module2tuples`` to generate a JSON file containing resource
+  definitions.
+
+.. code-block:: console
+
+    # in the root directory of the project (where the go.mod file lives)
+    $ go mod vendor
+    $ modules2tuple -spack -app_version=1.2.3 vendor/modules.txt > resources-1.2.3.json
+
+* Finally, copy that file into the Spack package directory and add an
+  `import_resources` statement that refers to it in the package
+  definition.
+
+.. code-block:: python
+
+        import_resources("resources-1.2.3.py", when="@1.2.3")
+
+^^^^^^^^^^^^^^^
+Advanced topics
+^^^^^^^^^^^^^^^
+
+Packaging projects that don't fit into either of the previous two
+categories is "left as an exercise for the reader".
+
+More seriously, there are two things to keep in mind:
+
+* the package should use a "run" dependency on go (unless something
+  wacky is happening at run time); and
+
+* you'll need to provide ``resource`` definitions for dependencies
+  (``modules2tuples`` might be helpful) and ensure that ``go`` does
+  not access the network while building (perhaps by invoking it with
+  the ``-mod=vendor`` flag).
+
+* ensure that the use of the ``GOFLAGS`` environment variable does not
+  conflict with attempts to set arguments on the command line.
```

Comment:
Can you add a final section with links to more documentation on building Go packages? That's what I did in the other build system docs.

Reply:
Will do.

Reply:
done.